HW #4 (GoF Patterns)

cholombytko opened the pull request:
This pull request all tasks related to Gof Patterns homework.
Repository already implement

cholombytko committed:
feat: add new api url

Add new free api url to get exchange rate

cholombytko committed:
Merge branch 'hw3/tests' of github.com:GenesisEducationKyiv/software-engineering-school-4-0-cholombytko into hw4/gof

cholombytko committed:
feat: add new currency endpoint

This commit contains new currency endpoint and some better naming
refactoring

cholombytko committed:
feat&refactor: add chain responsibility implementation and refactor

This commit contains:
-Chain responsibility implementation for handling several exchange rates
providers
-Refactor of whole module system using NestJS Dependency Injection
practices and custom providers
-Logger implementation for handlers

eluyema commented on src/rate/logging/logger.decorator.ts:
```diff
@@ -0,0 +1,27 @@
+import { Logger } from '@nestjs/common';
+
+export function Log(targetUrl: string) {
```
Maybe it would be more convenient to add something related to rate to the name ? For example: LogRateDecorator, LogRateRequest, etc

cholombytko replied:
Good catch!

eluyema commented on src/rate/providers/currency-api.service.ts:
```diff
@@ -0,0 +1,42 @@
+    return handler;
+  }
+
+  @Log(CURRENCY_API_URL)
```
Cool! I forgot about decorators in typescript/nest.js, and created just class for it )
I'll refactor my class to something like yours ))

eluyema commented:
Good job!
I have no questions to implementations of patterns

cholombytko committed:
refactor: better naming

This commit contains better naming for logging decorator

cholombytko committed:
refactor: delete unused logger

This commit contains deleting unused logger in nbuservice

MlPablo commented on src/rate/providers/nbu.service.ts:
```diff
@@ -0,0 +1,42 @@
+
+  constructor(private readonly httpService: HttpService) {}
+
+  setNext(handler: IHandler): IHandler {
+    this.nextHandler = handler;
+    return handler;
+  }
```
Це не зовсім паттерн Chain of responsibility. Тут в тебе кожен клієнт відповідає за знання Next та додає логіку передачі на наступний клієнт. Це має робити окремий об'єкт який буде відповідати за те щоб сетити та керувати потоком виконнаня по чейну.
т

cholombytko replied:
Дякую за ревʼю. Переробив імплементувавши Chain Manager. Посилання на коміт: 663aae0

MlPablo commented:
По іншому вцілому питань немаю (в кодстайл node.js не дивився)

cholombytko committed:
refactor: refactor of chain responsibility pattern

This commit contains severa; updates:
-Implementing chain manager
-Refactor logging decorator
-Refactor handlers

cholombytko committed:
fix: fix nbuhandler

This commit contains fix of nbuhandler by deleting error used in testing
chain workflow
